library/Translate/src/Juvix/Sexp/Structure.hs

```diff
@@ -0,0 +1,745 @@
+
+----------------------------------------
+-- Signature
+----------------------------------------
+
+nameSignature :: NameSymbol.T
+nameSignature = ":defsig"
+
+isSignature :: Sexp.T -> Bool
+isSignature (Sexp.Cons form _) = Sexp.isAtomNamed form nameSignature
+isSignature _ = False
+
+toSignature :: Sexp.T -> Maybe Signature
+toSignature form
+  | isSignature form =
+    case form of
+      _Signature Sexp.:> sexp1 Sexp.:> sexp2 Sexp.:> Sexp.Nil ->
+        Signature sexp1 sexp2 |> Just
+      _ ->
+        Nothing
+  | otherwise =
+    Nothing
+
+fromSignature :: Signature -> Sexp.T
+fromSignature (Signature sexp1 sexp2) =
+  Sexp.list [Sexp.atom nameSignature, sexp1, sexp2]
+
+----------------------------------------
+-- Let
+----------------------------------------
+
+nameLet :: NameSymbol.T
+nameLet = "let"
+
+isLet :: Sexp.T -> Bool
+isLet (Sexp.Cons form _) = Sexp.isAtomNamed form nameLet
+isLet _ = False
+
+toLet :: Sexp.T -> Maybe Let
+toLet form
+  | isLet form =
+    case form of
+      _Let Sexp.:> sexp1 Sexp.:> sexp2 Sexp.:> sexp3 Sexp.:> sexp4 Sexp.:> Sexp.Nil ->
+        Let sexp1 sexp2 sexp3 sexp4 |> Just
+      _ ->
+        Nothing
+  | otherwise =
+    Nothing
+
+fromLet :: Let -> Sexp.T
+fromLet (Let sexp1 sexp2 sexp3 sexp4) =
+  Sexp.list [Sexp.atom nameLet, sexp1, sexp2, sexp3, sexp4]
+
+----------------------------------------
+-- LetMatch
+----------------------------------------
+
+nameLetMatch :: NameSymbol.T
+nameLetMatch = ":let-match"
+
+isLetMatch :: Sexp.T -> Bool
+isLetMatch (Sexp.Cons form _) = Sexp.isAtomNamed form nameLetMatch
+isLetMatch _ = False
+
+toLetMatch :: Sexp.T -> Maybe LetMatch
+toLetMatch form
+  | isLetMatch form =
+    case form of
+      _LetMatch Sexp.:> sexp1 Sexp.:> argBodys2 Sexp.:> sexp3 Sexp.:> Sexp.Nil
+        | Just argBodys2 <- toArgBodys argBodys2 ->
+          LetMatch sexp1 argBodys2 sexp3 |> Just
+      _ ->
+        Nothing
+  | otherwise =
+    Nothing
+
+fromLetMatch :: LetMatch -> Sexp.T
+fromLetMatch (LetMatch sexp1 argBodys2 sexp3) =
+  Sexp.list [Sexp.atom nameLetMatch, sexp1, fromArgBodys argBodys2, sexp3]
+
+----------------------------------------
+-- PredAns
+----------------------------------------
+
+toPredAns :: Sexp.T -> Maybe PredAns
+toPredAns form =
+  case form of
+    sexp1 Sexp.:> sexp2 Sexp.:> Sexp.Nil ->
+      PredAns sexp1 sexp2 |> Just
+    _ ->
+      Nothing
+
+fromPredAns :: PredAns -> Sexp.T
+fromPredAns (PredAns sexp1 sexp2) =
+  Sexp.list [sexp1, sexp2]
+
+----------------------------------------
+-- Cond
+----------------------------------------
+
+nameCond :: NameSymbol.T
+nameCond = ":cond"
+
+isCond :: Sexp.T -> Bool
+isCond (Sexp.Cons form _) = Sexp.isAtomNamed form nameCond
+isCond _ = False
+
+toCond :: Sexp.T -> Maybe Cond
+toCond form
+  | isCond form =
+    case form of
+      _Cond Sexp.:> predAns1
+        | Just predAns1 <- toPredAns `fromStarList` predAns1 ->
+          Cond predAns1 |> Just
```

Comment:
Fascinated that your brain parses `Cond predAns1 |> Just` better that `Just $ Cond predAns1`